### PR TITLE
[CELEBORN-1318] Support celeborn http authentication

### PR DESCRIPTION
**Pull request by turboFei**
### What changes were proposed in this pull request?
Support celeborn master/worker http authentication.

### Why are the changes needed?
Authentication is needed for celeborn admin APIs.

### Does this PR introduce _any_ user-facing change?
Yes, introduce authentication related config items, but does not break the current behavior.

### How was this patch tested?

Added UT for BASIC and Bearer authentication.

### CODE REVIEW

**Comment by turboFei** [comment]
This is PR is almost ready. cc @RexXiong @SteNicholas 

I have not add spnego testing, because it will introduce many testing dependencies.

**Comment by turboFei** [comment]
also cc @pan3793

File: common/src/main/scala/org/apache/celeborn/common/authentication/PasswdAuthenticationProvider.scala [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+   * If a user is to be granted, return nothing/throw nothing.
+   * When a user is to be disallowed, throw an appropriate [[AuthenticationException]].
+   *
+   * @param user     The username received over the connection request
+   * @param password The password received over the connection request
+   *
+   * @throws AuthenticationException When a user is found to be invalid by the implementation
+   */
+  @throws[AuthenticationException]
+  def authenticate(user: String, password: String): Unit
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/authentication/PasswdAuthenticationProvider.scala:36]: return a `java.security.Principal` instead. you can refer to Trino's `io.trino.spi.security.HeaderAuthenticator`

File: common/src/main/scala/org/apache/celeborn/common/authentication/TokenAuthenticationProvider.scala [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+   * If the token is to be granted, return nothing/throw nothing.
+   * When the token is to be disallowed, throw an appropriate [[AuthenticationException]].
+   *
+   * @param token The token received over the connection request.
+   * @return The identifier associated with the token
+   *
+   * @throws AuthenticationException When the token is found to be invalid by the implementation
+   */
+  @throws[AuthenticationException]
+  def authenticate(token: String): String
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/authentication/TokenAuthenticationProvider.scala:36]: ditto, Principal

File: common/src/main/scala/org/apache/celeborn/common/authentication/PasswdAuthenticationProvider.scala [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.celeborn.common.authentication
+
+import javax.security.sasl.AuthenticationException
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/authentication/PasswdAuthenticationProvider.scala:20]: is this SASL related? looks like abuse

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 4]
```diff
@@ -2167,6 +2167,69 @@ object CelebornConf extends Logging {
...
+        " the client IP address, to get around this common issue, most load balancers or proxies" +
+        " offer the ability to record the real remote IP address in an HTTP header that will be" +
+        " added to the request for other devices to use. Note that, because the header value can" +
+        " be specified to any IP address, so it will not be used for authentication.")
+      .version("0.5.0")
+      .stringConf
+      .createWithDefault("X-Real-IP")
+
+  val MASTER_HTTP_AUTH_BASIC_PROVIDER_CLASS: ConfigEntry[String] =
+    buildConf("celeborn.master.http.auth.basic.provider.class")
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:2216]: let's remove `.class` and allow it to accept both the short name of the built-in implementations alias and user-provided 3-rd full class name
```suggestion
    buildConf("celeborn.master.http.auth.basic.provider")
```

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 5]
```diff
@@ -2167,6 +2167,69 @@ object CelebornConf extends Logging {
...
       .createWithDefaultString("30s")
 
+  val MASTER_HTTP_AUTH_SUPPORTED_SCHEMES: ConfigEntry[Seq[String]] =
+    buildConf("celeborn.master.http.auth.supportedSchemes")
+      .categories("master")
+      .version("0.5.0")
+      .doc("A comma-separated list of master http auth supported schemes." +
+        "<ul>" +
+        " <li>SPNEGO: Kerberos/GSSAPI authentication.</li>" +
+        " <li>BASIC: User-defined password authentication, anonymous by default.</li>" +
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:2177]:
```suggestion
        " <li>BASIC: User-defined password authentication, the concreted implementation is configurable via `celeborn.master.http.auth.basic.provider`.</li>" +
```

**Comment by turboFei** [comment]
just come back after paternity leave, will address the comments

**Comment by turboFei** [comment]
address all comments, cc @pan3793

**Comment by pan3793** [comment]
@turboFei welcome back, the changes seem partially applied to master/worker configurations, could you please revise again to align them? others lgtm.

**Comment by pan3793** [comment]
btw, main branch goes 0.6 now

**Comment by turboFei** [comment]
Thank you @pan3793 , aligned the config items between master and work, updated the conf version.

**Comment by turboFei** [comment]
cc  @pan3793

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 6]
```diff
@@ -2795,6 +2860,71 @@ object CelebornConf extends Logging {
...
 
+  val WORKER_HTTP_AUTH_SUPPORTED_SCHEMES: ConfigEntry[Seq[String]] =
+    buildConf("celeborn.worker.http.auth.supportedSchemes")
+      .categories("worker")
+      .version("0.6.0")
+      .doc("A comma-separated list of worker http auth supported schemes." +
+        "<ul>" +
+        " <li>SPNEGO: Kerberos/GSSAPI authentication.</li>" +
+        " <li>BASIC: User-defined password authentication, the concreted implementation is" +
+        " configurable via `celeborn.master.http.auth.basic.provider`.</li>" +
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:2871]: some places wrongly refer to `master` configurations
```suggestion
        " configurable via `celeborn.worker.http.auth.basic.provider`.</li>" +
```

**Review by pan3793 — APPROVED** [review]

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala (hunk 6 above)
Comment (turboFei) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala]: Done

**Comment by pan3793** [comment]
Thanks, merged to main for 0.6

**Comment by waitinfuture** [comment]
@turboFei welcome back!
